CUMULUS-2856: Allow removing execution fields

jennyhliu opened the pull request:
**Summary:** Summary of changes

Addresses [CUMULUS-2856: Update PUT /executions endpoint to allow removing fields](https://bugs.earthdata.nasa.gov/browse/CUMULUS-2856)

## Changes

* Update API/Message write logic to handle nulls as deletion in execution PUT/message write logic
* Update execution PUT endpoint to no longer respect message write constraints and update all values passed in

## PR Checklist

- [x] Update CHANGELOG
- [x] Unit tests
- [ ] Ad-hoc testing - Deploy changes and test manually
- [x] Integration tests

Jkovarik commented on packages/es-client/indexer.js:
```diff
@@ -317,7 +350,7 @@ async function upsertGranule({
 }, writeConstraints = true) {
   Object.keys(updates).forEach((key) => {
     if (updates[key] === null && granuleInvalidNullFields.includes(key)) {
-      throw new ValidationError(`Attempted DynamoDb write with invalid key ${key} set to null.  Please remove or change this field and retry`);
```
Oops!    Good catch.

Jkovarik commented on packages/es-client/tests/test-execution-upsert.js:
```diff
@@ -44,15 +45,17 @@ test.serial('upsertExecution writes new "running" execution', async (t) => {
 });
 
-test.serial('upsertExecution writes new "completed" execution', async (t) => {
+test.serial('upsertExecution writes new "completed" execution with null fields omitted', async (t) => {
```
We need a test that shows nullification happens on overwrite as well as new writes.

Jkovarik commented on packages/types/api/executions.d.ts:
```diff
@@ -7,9 +7,10 @@ export type ExecutionProcessingTimes = {
 
 export interface ExecutionRecord {
```
Should these types have a similar name, e.g. ApiExecutionRecord and ApiRecord for consistency with granules, pdrs, etc?

Jkovarik replied:
Resolved https://github.com/nasa/cumulus/pull/3186/commits/05e4e6bec64eec060eeb397702228c8613031cb9

Jkovarik commented on packages/types/api/executions.d.ts:
```diff
@@ -21,6 +22,25 @@ export interface ExecutionRecord {
   cumulusVersion?: string,
 }
+
+export interface ApiExecution {
```
Nit: Alpha sort

Jkovarik commented on CHANGELOG.md:
```diff
@@ -81,6 +81,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
     - Remove granule-csv endpoint
+  - **CUMULUS-2856**
+    - Update API/Message write logic to handle nulls as deletion in execution PUT/message write logic
+    - Update execution PUT endpoint to no longer respect message write constraints and update all values passed in
```
This might be a breaking change

Jkovarik replied:
Resolved in https://github.com/nasa/cumulus/pull/3186/commits/05e4e6bec64eec060eeb397702228c8613031cb9

Jkovarik reviewed (changes requested):
This looks good @jennyhliu , I just have a couple of suggestions/questions re: tests.   Let me know if you want to pair/talk through any of it!

jennyhliu commented:
@Jkovarik I have addressed your comments in [05e4e6b](https://github.com/nasa/cumulus/pull/3186/commits/05e4e6bec64eec060eeb397702228c8613031cb9) [a8c31f75](https://github.com/nasa/cumulus/pull/3186/commits/a8c31f75c73982d49a634a5342571ed3627ee901)

Jkovarik commented on packages/db/src/translate/executions.ts:
```diff
@@ -67,12 +71,36 @@ export const translatePostgresExecutionToApiExecution = async (
   return <ExecutionRecord>removeNilProperties(translatedRecord);
 };
 
+const returnNullOrUndefinedOrDate = (
```
@jennyhliu  I think this comment was missed.   It's not *super* important, but for maintainability we could probably put this in .... lib?     Mostly wanted to get your thoughts on the comment.

jennyhliu replied:
[d8bfba7](https://github.com/nasa/cumulus/pull/3186/commits/d8bfba7312a9909a872eda1be3158cbfc601a3e5)

Jkovarik commented on packages/api/lib/writeRecords/write-execution.js:
```diff
@@ -85,19 +85,21 @@ const buildExecutionRecord = ({
   const {
     apiRecord,
     esClient = await Search.es(),
+    writeConstraints,
```
Resolved https://github.com/nasa/cumulus/pull/3186/commits/05e4e6bec64eec060eeb397702228c8613031cb9

Jkovarik commented on packages/api/tests/endpoints/test-executions.js:
```diff
@@ -1424,6 +1426,146 @@ test.serial('PUT /executions updates the record as expected in PostgreSQL/Elasti
+    asyncOperationId: null,
+    collectionId: null,
+    cumulusVersion: null,
+    duration: 180.5,
```
Should duration be null here?

jennyhliu replied:
[d8bfba7](https://github.com/nasa/cumulus/pull/3186/commits/d8bfba7312a9909a872eda1be3158cbfc601a3e5)

Jkovarik reviewed (changes requested):
@jennyhliu  thanks for addressing the changes, just one more comment re: that unit and a comment/question. 

Thanks for your patience re: me getting back to you on this.

Jkovarik reviewed (approved):
Looks good!  Nice work!
